solver/llbsolver/ops/exec.go

```diff
@@ -214,149 +214,48 @@ func addDefaultEnvvar(env []string, k, v string) []string {
+		desc := fmt.Sprintf("mount %s from exec %s", m.Dest, strings.Join(e.op.Meta.Args, " "))
+		return e.cm.New(ctx, ref, g, cache.WithDescription(desc))
+	})
+	defer func() {
+		for i := len(p.Actives) - 1; i >= 0; i-- { // call in LIFO order
+			p.Actives[i].Release(context.TODO())
 		}
-
-		// if mount is based on input validate and load it
-		if m.Input != pb.Empty {
-			if int(m.Input) > len(inputs) {
-				return nil, errors.Errorf("missing input %d", m.Input)
+		for _, o := range p.OutputRefs {
+			if o.Ref != nil {
+				o.Ref.Release(context.TODO())
 			}
-			inp := inputs[int(m.Input)]
-			workerRef, ok := inp.Sys().(*worker.WorkerRef)
-			if !ok {
-				return nil, errors.Errorf("invalid reference for exec %T", inp.Sys())
-			}
-			ref = workerRef.ImmutableRef
-			mountable = ref
-		}
-
-		makeMutable := func(ref cache.ImmutableRef) (cache.MutableRef, error) {
-			desc := fmt.Sprintf("mount %s from exec %s", m.Dest, strings.Join(e.op.Meta.Args, " "))
-			return e.cm.New(ctx, ref, g, cache.WithDescription(desc))
 		}
-
-		switch m.MountType {
-		case pb.MountType_BIND:
-			// if mount creates an output
-			if m.Output != pb.SkipOutput {
-				// if it is readonly and not root then output is the input
-				if m.Readonly && ref != nil && m.Dest != pb.RootMount {
-					outputs = append(outputs, ref.Clone())
-				} else {
-					// otherwise output and mount is the mutable child
-					active, err := makeMutable(ref)
-					if err != nil {
-						return nil, err
-					}
-					outputs = append(outputs, active)
-					mountable = active
-				}
-			} else if (!m.Readonly || ref == nil) && m.Dest != pb.RootMount {
-				// this case is empty readonly scratch without output that is not really useful for anything but don't error
-				active, err := makeMutable(ref)
-				if err != nil {
-					return nil, err
+		if err != nil {
+			execInputs := make([]solver.Result, len(e.op.Mounts))
+			for i, m := range e.op.Mounts {
+				if m.Input == -1 {
+					continue
 				}
-				defer active.Release(context.TODO())
-				mountable = active
-			}
-
-		case pb.MountType_CACHE:
-			mRef, err := e.mm.MountableCache(ctx, m, ref, g)
-			if err != nil {
-				return nil, err
-			}
-			mountable = mRef
-			defer func() {
-				go mRef.Release(context.TODO())
-			}()
-			if m.Output != pb.SkipOutput && ref != nil {
-				outputs = append(outputs, ref.Clone())
-			}
-
-		case pb.MountType_TMPFS:
-			mountable = e.mm.MountableTmpFS()
-		case pb.MountType_SECRET:
-			var err error
-			mountable, err = e.mm.MountableSecret(ctx, m, g)
-			if err != nil {
-				return nil, err
-			}
-			if mountable == nil {
-				continue
-			}
-		case pb.MountType_SSH:
-			var err error
-			mountable, err = e.mm.MountableSSH(ctx, m, g)
-			if err != nil {
-				return nil, err
+				execInputs[i] = inputs[m.Input]
 			}
-			if mountable == nil {
-				continue
+			execOutputs := make([]solver.Result, len(e.op.Mounts))
+			for i, res := range results {
```

Comment:
I still don't quite understand how this works. We are mapping only mounts that had set output index, instead of all the mounts(that would then point to either mutable or same input if readonly). I think the test works because something always fills up output index there, even if it is unused.

Reply:
Addressed here: https://github.com/moby/buildkit/pull/1732/commits/1240dd7795823218c049782ce0302776a69e0739